frame/dutch-auction/src/lib.rs

```diff
@@ -110,204 +57,64 @@ pub mod pallet {
 	pub type OrdersIndex<T: Config> = StorageValue<_, T::OrderId, ValueQuery>;
 
-	impl<T: Config + DeFiComposableConfig> DutchAuction for Pallet<T> {
-		type AccountId = T::AccountId;
+	#[derive(Encode, Decode, Default, TypeInfo)]
+	pub struct SellOrder<OrderId, AssetId, Balance, AccountId> {
+		pub id: OrderId,
+		pub order: Sell<AssetId, Balance>,
+		// allow to sell by parts like Order Book does, sorted by best proposition
+		pub takes: Vec<TakeBy<AccountId, Balance>>,
+	}
 
+	#[pallet::storage]
+	#[pallet::getter(fn buys)]
+	pub type SellOrders<T: Config> =
+		StorageMap<_, Twox64Concat, OrderIdOf<T>, SellOrderOf<T>, OptionQuery>;
+
+	impl<T: Config + DeFiComposableConfig> DeFiEngine for Pallet<T> {
 		type AssetId = T::AssetId;
 
 		type Balance = T::Balance;
 
-		type OrderId = T::OrderId;
-
-		type Orderbook = T::Orderbook;
-
-		type GroupId = T::GroupId;
+		type AccountId = T::AccountId;
+	}
 
-		#[allow(clippy::type_complexity)]
-		type Order = Order<OrderIdOf<T>, T::AccountId, T::AssetId, T::Balance, T::GroupId>;
+	impl<T: Config + DeFiComposableConfig> SellEngine<AuctionStepFunction> for Pallet<T> {
+		type OrderId = T::OrderId;
 
-		fn start(
-			account_id: &Self::AccountId,
-			source_asset_id: Self::AssetId,
-			source_account: &Self::AccountId,
-			target_asset_id: Self::AssetId,
-			target_account: &Self::AccountId,
-			total_amount: Self::Balance,
-			price: PriceStructure<Self::GroupId, Self::Balance>,
-			function: AuctionStepFunction,
+		fn ask(
+			_from_to: &Self::AccountId,
+			_order: composable_traits::defi::Sell<Self::AssetId, Self::Balance>,
+			_base_amount: Self::Balance,
+			_configuration: AuctionStepFunction,
 		) -> Result<Self::OrderId, DispatchError> {
-			// TODO: with remote foreign chain DEX it can pass several blocks before we get on DEX.
-			// so somehow need to lock (transfer) currency before foreign transactions settles
-			ensure!(
-				<T::Currency as Inspect<T::AccountId>>::can_withdraw(
-					source_asset_id,
-					account_id,
-					total_amount
-				)
-				.into_result()
-				.is_ok(),
-				Error::<T>::CannotWithdrawAmountEqualToDesiredAuction
-			);
-
-			// because dex call is in "other transaction" and same block can have 2 starts each
-			// passing check, but failing during dex call.
-			let order_id: T::OrderId = OrdersIndex::<T>::get();
-			OrdersIndex::<T>::set(order_id.next());
-
-			let order = Order {
-				started: T::UnixTime::now().as_secs(),
-				function,
-				account_id: account_id.clone(),
-				source_asset_id,
-				source_account: source_account.clone(),
-				target_asset_id,
-				target_account: target_account.clone(),
-				source_total_amount: total_amount,
-				source_initial_price: price,
-				state: AuctionState::AuctionStarted,
-			};
-			Orders::<T>::insert(order_id.clone(), order);
-			Self::deposit_event(Event::<T>::AuctionWasStarted { order_id: order_id.clone() });
-			Ok(order_id)
+			Self::deposit_event(Event::OrderAdded { order_id: <_>::default() });
+			todo!()
```

Comment:
This is for next PR?

Reply:
yes. i am working on implementation now.